### PR TITLE
added ability to handle parent classes for PropertyNormalizer

### PR DESCRIPTION
**Pull request by ivoba**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes
| Fixed tickets | #24152 
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

This adds the ability for PropertyNormalizer to normalize/denormalize properties from parent classes.

### CODE REVIEW

File: src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php [hunk 1]
```diff
@@ -84,6 +84,21 @@ protected function isAllowedAttribute($classOrObject, $attribute, $format = null
                 return false;
             }
         } catch (\ReflectionException $reflectionException) {
+            $className = is_string($classOrObject) ? $classOrObject : get_class($classOrObject);
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: The first check (line 82) can also be in the while right?

File: src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php [hunk 2]
```diff
@@ -84,6 +84,21 @@ protected function isAllowedAttribute($classOrObject, $attribute, $format = null
...
+            $class = new \ReflectionClass($className);
+
+            while ($parent = $class->getParentClass()) {
+                try {
+                    $reflectionProperty = new \ReflectionProperty($parent->getName(), $attribute);
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: Static methods must be excluded.

Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: As you have an instance of `ReflectionClass` available, maybe can you just call `getProperty`?

File: src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php [hunk 3]
```diff
@@ -84,6 +84,21 @@ protected function isAllowedAttribute($classOrObject, $attribute, $format = null
...
+                    if ($reflectionProperty) {
+                        return true;
+                    }
+                } catch (\Exception $e) {
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: Only `ReflectionException` should be catched.

File: src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php [hunk 4]
```diff
@@ -117,7 +134,13 @@ protected function getAttributeValue($object, $attribute, $format = null, array
         try {
             $reflectionProperty = new \ReflectionProperty(get_class($object), $attribute);
         } catch (\ReflectionException $reflectionException) {
-            return;
+            $class = get_parent_class($object);
+            do {
+                $reflectionProperty = new \ReflectionProperty(get_parent_class($object), $attribute);
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: Exceptions aren't caught here.

File: src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php [hunk 5]
```diff
@@ -117,7 +134,13 @@ protected function getAttributeValue($object, $attribute, $format = null, array
         try {
             $reflectionProperty = new \ReflectionProperty(get_class($object), $attribute);
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: Same here, should be in the `do/while`

File: src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php [hunk 6]
```diff
@@ -117,7 +134,13 @@ protected function getAttributeValue($object, $attribute, $format = null, array
...
         } catch (\ReflectionException $reflectionException) {
-            return;
+            $class = get_parent_class($object);
+            do {
+                $reflectionProperty = new \ReflectionProperty(get_parent_class($object), $attribute);
+                if (null === $reflectionProperty) {
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: How can it be `null`?

File: src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php [hunk 7]
```diff
@@ -136,7 +159,13 @@ protected function setAttributeValue($object, $attribute, $value, $format = null
         try {
             $reflectionProperty = new \ReflectionProperty(get_class($object), $attribute);
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: Same here.

**Review by dunglas — CHANGES_REQUESTED** [review]

**Comment by kbond** [comment]
@ivoba I suggest adding a private method `getReflectionProperty` to reduce the duplication, ie:

```php
private function getReflectionProperty($classOrObject, $attribute)
{
    $class = is_string($classOrObject) ? $classOrObject : get_class($classOrObject);

    try {
        return new \ReflectionProperty($class, $attribute);
    } catch (\ReflectionException $e) {
        $class = new \ReflectionClass($class);

        while ($parent = $class->getParentClass()) {
            try {
                return new \ReflectionProperty($parent->getName(), $attribute);
            } catch (\ReflectionException $e) {
                $class = $parent;
            }
        }

        throw $e;
    }
}
```

**Comment by ivoba** [comment]
i updated the PR and took @kbond suggestion into account.
@dunglas pls review if the changes fit your comments.

i have no idea why the test in PHP7.1 fail, looks not really related??

File: src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php [hunk 8]
```diff
@@ -150,4 +152,33 @@ protected function setAttributeValue($object, $attribute, $value, $format = null
...
+
+    /**
+     * @param string|object $classOrObject
+     * @param string        $attribute
+     *
+     * @return \ReflectionProperty
+     *
+     * @throws \ReflectionException
+     */
+    private function getReflectionProperty($classOrObject, $attribute)
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: What about:

```
    private function getReflectionProperty($classOrObject, $attribute)
    {
        $class = is_string($classOrObject) ? $classOrObject : get_class($classOrObject);
        while (true) {
            try {
                return isset($reflectionClass) ? $reflectionClass->getProperty($attribute) : new \ReflectionProperty($class, $attribute);
            } catch (\ReflectionException $e) {
                $reflectionClass = new \ReflectionClass($class);
                if (!$reflectionClass->getParentClass()) {
                    throw $e;
                }
            }
        }
    }
```

Comment (ivoba) [reply src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: This runs into an endless loop.

Comment (ivoba) [reply src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: Would work with this:

    ... 
    $reflectionClass = new \ReflectionClass($class);
    $reflectionClass = $reflectionClass->getParentClass();
    if ($reflectionClass === false) {
        throw $e;
    }
    ...

Comment (dunglas) [reply src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: I've updated my snippet, it should work now

Comment (ivoba) [reply src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: i updated the PR with the snippet included

**Review by dunglas — APPROVED** [review]

**Review by Simperfit — APPROVED** [review]

File: src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php [hunk 9]
```diff
@@ -150,4 +152,27 @@ protected function setAttributeValue($object, $attribute, $value, $format = null
...
+    {
+        $class = is_string($classOrObject) ? $classOrObject : get_class($classOrObject);
+        while (true) {
+            try {
+                return isset($reflectionClass) ? $reflectionClass->getProperty($attribute) : new \ReflectionProperty($class, $attribute);
+            } catch (\ReflectionException $e) {
+                $reflectionClass = new \ReflectionClass($class);
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: Wouldn't it be better to initialize this outside the loop?

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php (hunk 9 above)
Comment (fabpot) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: Looks wrong to me as you are always overriding `$reflectionClass` with `$class`. This statement should only be executed if `$reflectionClass` is not set. I would add a test with a more complex hierarchy: class < parent < parent. Current code would loop forever.

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by ivoba** [comment]
I added a new Dummy class with 3 level inheritance and applied suggested fixes.

File: src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php [hunk 10]
```diff
@@ -150,4 +152,27 @@ protected function setAttributeValue($object, $attribute, $value, $format = null
...
+     * @param string|object $classOrObject
+     * @param string        $attribute
+     *
+     * @return \ReflectionProperty
+     *
+     * @throws \ReflectionException
+     */
+    private function getReflectionProperty($classOrObject, $attribute)
+    {
+        $class = is_string($classOrObject) ? $classOrObject : get_class($classOrObject);
```
Comment (fabpot) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: This is not needed as the `ReflectionClass` constructor takes either a class or an object as an argument.

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by ivoba** [comment]
removed mentioned line. 
fabbot fails for License Headers but diff is empty. probably a fabbot bug?

**Comment by xabbuh** [comment]
your `GroupDummyChild.php` fixtures file is missing the license header

**Comment by ivoba** [comment]
ok thx, licence header is added

**Review by xabbuh — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @ivoba.

**Pull request closed without merging**